go/vt/mysqlctl/backupstorage/params.go

```diff
@@ -0,0 +1,47 @@
+/*
+Copyright 2022 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package backupstorage
+
+import (
+	"vitess.io/vitess/go/vt/logutil"
+	"vitess.io/vitess/go/vt/mysqlctl/backupstats"
+	logutilpb "vitess.io/vitess/go/vt/proto/logutil"
+)
+
+// Params contains common utilities that may be used by BackupStorage and FileHandle components.
+// This gives out-of-three backup/restore plugins mechanisms to log and generate metrics,
+// while keeping policies (metric names and labels, logging destination, etc.)
+// in the control of in-tree components and Vitess users.
+type Params struct {
+	Logger logutil.Logger
+	Stats  backupstats.Stats
+}
+
+// Parameterizer gives BackupStorage components a way to opt in to receiving
+// Params without breaking the current API.
+type Parameterizer interface {
+	WithParams(Params) BackupStorage
+}
```

Comment:
> What happens when we add more params?

Hm I'm not sure. I think if we add new fields to the `Params` struct it shouldn't cause any compiler errors? Renaming and removing params would cause errors of course. 

We wouldn't be able to change the `Parameterizer` interface, which makes it kind of brittle.

> We have to weigh whether it is better to allow them to continue without metrics, or to force them to implement metrics. I'd say we force them to implement metrics

If we are OK forcing people to update their out-of-tree plugins if/when this is released I am fine with that too!

I guess the question then becomes, if we force the change now, does it make sense to add `WithParams` to the `BackupStorage` interface, or do something else instead? 

To me using `WithParams` feels a bit awkward if are making a breaking change. 

Here are some thoughts about other options:

### Add `Params` to  `BackupStorage` interface

This isn't elegant, but it's straight-forward. 

Biggest downside in my mind is it will require the most code changes to backup storage implementations to pass `Params` downwards to `BackupHandle`, but that can be done in multiple PRs to make it manageable.

Changes would look like this:

```
type BackupStorage interface {
        - ListBackups(ctx context.Context, dir string) ([]BackupHandle, error)
	+ ListBackups(ctx context.Context, params Params, dir string) ([]BackupHandle, error)
	- StartBackup(ctx context.Context, dir, name string) (BackupHandle, error)
	+ StartBackup(ctx context.Context, params Params, dir, name string) (BackupHandle, error)
	- RemoveBackup(ctx context.Context, dir, name string) error
	+ RemoveBackup(ctx context.Context, params Params, dir, name string) error
	- Close() error
	+ Close(params Params) error
}
```

### Change `Context` to `BackupStorageContext`

Introduce a new object `BackupContext` that extends `Context`. Same as last option, just different aesthetic appeal.

```
+ type BackupStorageContext interface {
+        context.Context
+        Logger() logutil.Logger
+        Stats() backupstats.Stats
+ }
+
type BackupStorage interface {
        - ListBackups(ctx context.Context, dir string) ([]BackupHandle, error)
	+ ListBackups(ctx BackupStorageContext, dir string) ([]BackupHandle, error)
	- StartBackup(ctx context.Context, dir, name string) (BackupHandle, error)
	+ StartBackup(ctx BackupStorageContext, dir, name string) (BackupHandle, error)
	- RemoveBackup(ctx context.Context, dir, name string) error
	+ RemoveBackup(ctx BackupStorageContext, dir, name string) error
	- Close() error
	+ Close(ctx BackupStorageContext) error
}
```

### Introduce `BackupStorageFactory`

Change `map[string]BackupStorage` to `map[string]BackupStorageFactory`.

```
+type BackupStorageFactory = func(Params) BackupStorage
+
- var BackupStorageMap = make(map[string]BackupStorage)
+ var BackupStorageMap = make(map[string]BackupStorageFactory)
```

Add factories to each backup storage implementation. I do like that doing this attaches `Params` to `BackupStorage` and should mean less work to make `Params` available to `BackupHandle`, but overall I think the little amount of effort we save doesn't justify this kind of fancy approach.